onadata/apps/logger/management/commands/populate_media_file_basename.py

```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python
+# vim: ai ts=4 sts=4 et sw=4 coding=utf-8
+from django.conf import settings
+from django.db import connection
+from django.db.models import Q, Func
+from django.db.models.functions import Substr
+from django.core.management.base import BaseCommand, CommandError
+from django.utils.translation import ugettext as _, ugettext_lazy
+from optparse import make_option
+
+from onadata.apps.logger.models.attachment import Attachment
+
+
+class SubstrFromPattern(Func):
+    function = "SUBSTRING"
+    template = "%(function)s(%(expressions)s from '%(pattern)s')"
+
+
+class Command(BaseCommand):
+
+    help = ugettext_lazy("Updates indexed field `media_file_basename` which is empty or null")
+    option_list = BaseCommand.option_list + (
+        make_option(
+            '--batchsize',
+            type='int',
+            default=100,
+            help=ugettext_lazy("Number of records to process per query")),)
+
+    def handle(self, *args, **kwargs):
+        batchsize = kwargs.get("batchsize", 100)
```

Comment:
Is this needed with the `make_option` default? Also, since you don't load anything from the `Attachment` records into memory except for their PKs, I think that the default batch size could be a lot larger. When I ran the SQL that this replaces, I didn't do it in batches at all. (I do understand that having a progress indicator is nice.)

Reply:
Right. I've started from another command I created a while ago. I thought we wanted everything to be run by batch.

Reply:
There's nothing wrong with doing this by batch, but it's good to understand why chunking querysets is necessary sometimes (and when it might be unneeded).

Recall that problems arise when psycopg2 reads all rows of a result set from the Postgres server:
> So, let’s say you for some reason want to process 100 million rows. You know for sure that you won’t be able to hold all 100 million Model objects in memory, so you dutifully do:
> ```python
> qs = GiganticTableModel.objects.all()
> for giant in qs.iterator():
>     # And BANG, you get an out of memory exception right here.
> ```
> But what happened? Why did you run out of memory before you even saw a single object out of .iterator()?
> Let’s take a moment and trace down the code path that gets executed here:
> 
> * Creating the QuerySet doesn’t touch the database at all, as noted above.
> * After a certain amount of fussing around, .iterator() calls the underlying backend query machinery to perform the query.
> * The backend machinery executes the query, and creates an iterator over the resulting rows. That iterator (in this case) grabs a chunk of rows at a time using .fetchmany(), and returns them one at a time. (At it happens, that chunk is hardcoded at 100 rows.)
> * That iterator is called by the actual iterator returned by .iterator(), so the iteration (pfew!) proceeds as: Call to get a row (which refills if the last grab of 100 is exhausted), create a new object, and return it up. Create an object from that row, return it to the caller.
> 
> So, why are we getting an out of memory condition? Even though there are 100 million rows in the result, there should only be 100 in memory at any one time, right?
> 
> Sadly, wrong. At the moment that the backend machinery executes the query, all 100 million rows are returned by the database server at once.
> 
> To quote the psycopg2 documentation:
> 
> > When a database query is executed, the Psycopg cursor usually fetches all the records returned by the backend, transferring them to the client process. If the query returned an huge amount of data, a proportionally large amount of memory will be allocated by the client.
> 
> This is true even if you do a .fetchone() or .fetchmany(), not just a .fetchall(). And there’s no way, while staying entirely within the standard Django QuerySet machinery, to change this behavior. (https://thebuild.com/blog/2010/12/13/very-large-result-sets-in-django-using-postgresql/)

If each row merely contains the primary key (accomplished with `values_list("id", flat=True)`), it's unlikely that we'll run out of memory even if all rows are read at once by psycopg2.

PS: Django 1.11+ [includes server-side database cursors](https://github.com/django/django/commit/f3b7c059367a4e82bbfc7e4f0d42b10975e79f0c), rendering chunking completely unnecessary :sunglasses: